De-experimentalize st.fragment

vdonato opened the pull request:
We've had a few bugs filed and additional features requested for `@st.experimental_fragment`, but
for the most part, the feature seems to work quite well / we haven't seen much evidence that any drastic
changes to the API are required. While we may still make some tweaks to the feature going forward, it
does seem ok at this point to drop the `experimental_` prefix for the decorator to make it an officially
supported part of the Streamlit API!

This PR de-experimentalizes the `@st.fragment` decorator, and in the process it makes the following
changes to the feature.

* `st.rerun(scope="fragment")` can now be called from within a fragment to rerun only that fragment.
   We also added `st.rerun(scope="app")`, which functions the same way as calling `st.rerun()` without
   an argument does today.
* fragments can now be nested, which includes allowing developers to call fragments from within a dialog.
   Note: it is still prohibited to create a dialog from within a dialog. 
* it is now explicitly prohibited for a fragment to write a widget outside of the fragment.
* various bugfixes, including:
   * Exceptions raised from within a fragment can no longer interrupt other "simultaneously" running
      fragments.
   * Exceptions raised from within a fragment now print their error messages inside of the fragment's
      container rather than "globally" in the app.
   * We no longer incorrectly reraise `RuntimeError`s when a `KeyError` is raised within a fragment.
      * This may not completely get rid of instances where we're seeing users run into `RuntimeError`s
         with the message `"Could not find fragment with id <fragment_id>"`, but I do expect a substantial
         number of them to go away.

Closes #8635
Closes #8494
Closes #8591

vdonato committed:
Don't immediately remove fragments added by callbacks (#8916)

The reason we were seeing the bug in #8591 is due to the incorrect assumption that we can always
totally clear out a session's fragment storage before a full script run as fragments will immediately
be repopulated anyway. This isn't quite true as a fragment can be added by a callback, and in this
case clearing out `FragmentStorage` before the full script run will lose it.

To fix this, we keep track of the fragments added throughout a script run, and on a full script run
we don't clear out fragments that were just added.

vdonato committed:
Set ctx.current_fragment_id in both full script and fragment runs (#8931)

We previously used `ctx.current_fragment_id` to determine whether a script runner execution is a full
script run or a fragment rerun. We later switched to using `ctx.fragment_ids_this_run` to support
multiple fragments being run from a single script runner but still had the script runner be responsible
for setting `ctx.current_fragment_id` outside of the fragment's outer function.

It turns out that having the fragment's outer function be reponsible for this fixes the issue we were having
with nested fragments not working correctly, so this PR makes the simple required change to fix nested
fragments/dialogs.

Closes #8635

vdonato committed:
Remove fragment_ids_this_run from script run context (#8953)

I realized while working on the `st.rerun(scope="fragment")` feature that `ctx. fragment_ids_this_run`
should always be `rerun_data.fragment_id_queue` in set form. We still want to keep the distinction
around since the frontend isn't aware of the order fragments will be run, but we should simplify things
to not needlessly copy around the same information in a slightly different representation.

vdonato committed:
Deprecate @st.experimental_fragment and add @st.fragment (#8944)

This PR is being made a bit early since there's still a bit of work to do before we're truly ready to
de-experimentalize this feature, but I was making some other purely mechanical changes and
wanted to avoid starting on something more involved. We're also merging this into a feature
branch, so not quite being done with all of our planned changes for GA
isn't particularly problematic.

vdonato committed:
Implement st.rerun(scope="fragment") (#8979)

This PR adds the new `scope` kwarg to the `st.rerun()` function. With `scope="app"` (the default),
`st.rerun()` work as before. If you set `scope="fragment"` within a fragment, however, it's now possible
to specify that only the currently running fragment reruns.

vdonato committed:
Throw/expect more specific error when hitting KeyErrors in FragmentStorage methods (#9011)

A source of many (but potentially not all) of the reports in #8494 involve a silly mistake that I made
where we're catching all `KeyError`s that may be thrown from a fragment and interpreting them as
coming from when we attempted to fetch the fragment from our `FragmentStorage`. Of course, they
could also be coming from the fragment function itself, and in these cases we're currently raising an
error with a misleading / incorrect / opaque message.

This PR fixes this issue by throwing and expecting a more specific subclass of `KeyError` in
`FragmentStorage` operations

vdonato commented:
Going to temporarily add the `do-not-merge` label to this one for now until this feature goes through QA.

vdonato committed:
Remove mistakenly added whitespace

vdonato committed:
Remove experimental_ prefix from some decorator usage

vdonato committed:
Fix e2e test that miraculously still passed

vdonato commented on lib/streamlit/commands/execution_control.py:
```diff
@@ -54,22 +59,76 @@ def stop() -> NoReturn:  # type: ignore[misc]
 
     When ``st.rerun()`` is called, the script is halted - no more statements will
-    be run, and the script will be queued to re-run from the top.
+    be run, and the script will be queued to re-run.
+
+    Parameters
+    ----------
+    scope : Literal["app", "fragment"]
+        Specifies what part of the app should rerun. Setting scope="app" reruns the
+        full script; scope="fragment" limits the rerun to only the fragment from which
+        this function is called. If unspecified, defaults to "app".
+
+        Note that setting scope="fragment" is only valid
+            * inside of a fragment
+            * *not* during a full script run.
+        Passing this argument to ``st.rerun()`` from outside of a fragment or within a
+        fragment but when the full script is running raises a ``StreamlitAPIException``.
```
@sfc-gh-dmatthews heads up this docstring change is probably worth reviewing / editing

robingie replied:
While reading this, further extending the scope parameters may be beneficial. In certain use cases, it could be helpful to reference a specific named fragment that needs to be rerun.

For example:

st.rerun(scope='result_fragment') - named fragment
or
st.rerun(scope='parent') # referring to parent or sibling fragments

For example, this would be advantageous if you have a filter control in one fragment that queues data and displays the result in another fragment. This approach would be particularly useful for larger sites with multiple fragments, especially if some fragments have slightly longer execution times.

vdonato replied:
Thanks for the input @robingie. Something similar was brought up as a useful addition in #8511.

I think this is a feature that we may plan to add after removing the `experimental_` feature from the fragment decorator (which is already scheduled to happen next release), but it'll require some thought on our part to ensure we have the exact desired API / behavior worked out.

MarkDw123 commented:
> We've had a few bugs filed and additional features requested for `@st.experimental_fragment`, but for the most part, the feature seems to work quite well / we haven't seen much evidence that any drastic changes to the API are required. While we may still make some tweaks to the feature going forward, it does seem ok at this point to drop the `experimental_` prefix for the decorator to make it an officially supported part of the Streamlit API!
> 
> This PR de-experimentalizes the `@st.fragment` decorator, and in the process it makes the following changes to the feature.
> 
>     * `st.rerun(scope="fragment")` can now be called from within a fragment to rerun only that fragment.
>       We also added `st.rerun(scope="app")`, which functions the same way as calling `st.rerun()` without
>       an argument does today.
> 
>     * fragments can now be nested, which includes allowing developers to call fragments from within a dialog.
>       Note: it is still prohibited to create a dialog from within a dialog.
> 
>     * it is now explicitly prohibited for a fragment to write a widget outside of the fragment.
> 
>     * various bugfixes, including:
>       
>       * Exceptions raised from within a fragment can no longer interrupt other "simultaneously" running
>         fragments.
>       * Exceptions raised from within a fragment now print their error messages inside of the fragment's
>         container rather than "globally" in the app.
>       * We no longer incorrectly reraise `RuntimeError`s when a `KeyError` is raised within a fragment.
>         
>         * This may not completely get rid of instances where we're seeing users run into `RuntimeError`s
>           with the message `"Could not find fragment with id <fragment_id>"`, but I do expect a substantial
>           number of them to go away.
> 
> 
> Closes #8635 Closes #8494

Good morning. Is it correct that fragments with timer don't work in an st.expander? I am using an st aggrid with timer in a section. It works fine outside an expander, but fails inside one.

raethlein commented:
@vdonato I think you can also add that calling fragments from callbacks works now which closes #8591 🙂

raethlein reviewed (approved):
Let's ship it 🚀

vdonato commented:
> Good morning. Is it correct that fragments with timer don't work in an st.expander? I am using an st aggrid with timer in a section. It works fine outside an expander, but fails inside one.

Hi @MarkDw123,

Please file an issue with a minimal code example that reproduces the problem if you want to report a potential bug.

vdonato committed:
Fix issue with button clicks for enqueued fragments being dropped (#9062)

The reason we end up dropping trigger values in this case is that `coalesce_widget_states` fails
to take into account that the `WidgetStateManager` on the frontend immediately clears trigger
values after sending the `BackMsg` informing the server that widget values have changed.

This fixes the incorrect assumption in the function so that we still keep the old trigger value if we
can't find a widget with the same ID in `new_state` (which will essentially always be the case).

raethlein committed:
Always add fragment_ids to set (#9069)

## Describe your changes

Fix two issues:
- the `current_fragment_delta_path` could have been taken from the wrong
snapshot
- always add the `current_fragment_id` to the set of new fragment ids.
This prevents an issue where a fragment rerun is triggered during full
app execution.

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Add e2e test to ensure that the `current_fragment_id` error does not
happen
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

raethlein committed:
Fix duplicate dialog fragment id (#9066)

## Describe your changes

Right now, dialogs within an app always generate the same fragment id.
This is problematic and can lead to mixing contents of different
dialogs. Add a new parameter that is used by the dialog decorator to
pass the actually wrapped function name to the fragment in order to
generate the hash.
This behavior started to surface in the `feature/st.fragment` branch; prior to that we overwrote
fragments with the same id in the storage and since we only allow a single dialog at any given time, this issue stayed unrecognized.

## GitHub Issue Link (if applicable)

## Testing Plan

- Unit Tests (JS and/or Python)
- Unit tests seem difficult to add for this because if we use the
`DeltaGeneratorTestCase`, the dialogs have different delta paths when
opened simultaneously versus the same delta path in a real app after a
rerun happened. Hence, I figured the e2e test resembles the chain of
sequence best.
- E2E Tests
- Add an e2e test that triggers the sequence of action that leads to the
bug

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

raethlein committed:
Fix fragment exception impacts queued fragments (#9041)

## Describe your changes

In the current state, an exception in a fragment will cause queued
fragments to stop from running. But instead we want queued fragments to
still run even if a previous fragment raised an exception (even if the
error happens in a full app run; the error will be displayed but the app
will continue to run).

To do so, we add except blocks into the loop and continue running the
other fragments in the queue.

Also, we want to stop app execution when a fragment raises an error
during a full app run. This is why the `exec_func_with_error_handling`
call is removed in `fragment.py`.

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.
